### PR TITLE
Find deploy by ID

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Provides a way of searching for a deploy using it's id instead of inspecting the tuplespace

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-3456

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala [hunk 1]
```diff
@@ -70,6 +72,13 @@ object DeployRuntime {
       }.map(kp("")).value
     }
 
+  def findDeploy[F[_]: Functor: Sync: Time: DeployService](
+      deployId: String
+  ): F[Unit] =
+    gracefulExit(
+      DeployService[F].findDeploy(FindDeployQuery(Base16.unsafeDecode(deployId).toByteString))
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala]: I'm not sure it's safe to do Base16.unsafeDecode here. 

Who calls this function? Is it possible that the deployId is invalid Base16?

Maybe you should move the decoding to DeployService.findDeploy and deal with the error there?

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 2]
```diff
@@ -490,6 +490,17 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
   addSubcommand(dataAtName)
   addSubcommand(contAtName)
 
+  val findDeploy = new Subcommand("find-deploy") {
+    descr(
+      "Searches for a block containing the deploy with provided id."
+    )
+    val deployId = opt[String](
```
Comment (odeac) [inline node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: ...or maybe you should parse the deployId here using the coop.rchain.node.configuration.commandline.Base16Converter

File: casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala]: Ok, I'll add validation here

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala (hunk 2 above)
Comment (lukasz-golebiewski) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: Done

File: casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala [hunk 3]
```diff
@@ -70,6 +72,13 @@ object DeployRuntime {
       }.map(kp("")).value
     }
 
+  def findDeploy[F[_]: Functor: Sync: Time: DeployService](
+      deployId: Array[Byte]
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/util/comm/DeployRuntime.scala]: You should use consistently either DeployId type of Array[Byte] type

File: node/src/main/scala/coop/rchain/node/configuration/model.scala [hunk 4]
```diff
@@ -66,6 +66,7 @@ final case class Deploy(
     privateKey: Option[PrivateKey],
     location: String
 ) extends Command
+final case class FindDeploy(id: Array[Byte])                               extends Command
```
Comment (odeac) [inline node/src/main/scala/coop/rchain/node/configuration/model.scala]: DeployId ?

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 5]
```diff
@@ -490,6 +490,17 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
   addSubcommand(dataAtName)
   addSubcommand(contAtName)
 
+  val findDeploy = new Subcommand("find-deploy") {
+    descr(
+      "Searches for a block containing the deploy with provided id."
+    )
+    val deployId = opt[Array[Byte]](
```
Comment (odeac) [inline node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: DeployId instad of Array[Byte]?

File: node/src/main/scala/coop/rchain/node/configuration/model.scala (hunk 4 above)
Comment (lukasz-golebiewski) [reply node/src/main/scala/coop/rchain/node/configuration/model.scala]: `model` doesn't know about `casper`, and that's where `DeployId` is defined

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 6]
```diff
@@ -490,6 +490,18 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
   addSubcommand(dataAtName)
   addSubcommand(contAtName)
 
+  val findDeploy = new Subcommand("find-deploy") {
+    descr(
+      "Searches for a block containing the deploy with provided id."
+    )
+    implicit val converter: ValueConverter[Array[Byte]] = Base16Converter
+    val deployId = opt[Array[Byte]](
+      descr = "Id of the depoy.",
```
[the remark this answers is not in the document]
Comment (lukasz-golebiewski) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: You're commenting on outdated code

File: node/src/main/scala/coop/rchain/node/configuration/model.scala (hunk 4 above)
Comment (odeac) [reply node/src/main/scala/coop/rchain/node/configuration/model.scala]: ok

**Comment by lukasz-golebiewski** [comment]
bors try

**Comment by lukasz-golebiewski** [comment]
bors retry

File: casper/src/test/scala/coop/rchain/casper/MultiParentCasperDeploySpec.scala [hunk 7]
```diff
@@ -26,17 +26,18 @@ class MultiParentCasperDeploySpec extends FlatSpec with Matchers with Inspectors
     buildGenesisParameters(4, createBonds(validatorPks))
   )
 
-  "MultiParentCasper" should "accept deploys" in effectTest {
+  "MultiParentCasper" should "accept a deploy and return it's id" in effectTest {
```
Comment (dckc) [inline casper/src/test/scala/coop/rchain/casper/MultiParentCasperDeploySpec.scala]: s/it's/its/

it's = it is

**Pull request merged**